constraint-solver/src/quadratic_symbolic_expression.rs

```diff
@@ -846,20 +840,39 @@ mod tests {
         let b = Qse::from_known_symbol("B", RangeConstraint::default());
         let mut t: Qse = (x * a + y) * b;
         assert_eq!(t.to_string(), "(A * B) * X + B * Y");
-        t.apply_update(&VariableUpdate {
-            variable: "B",
-            known: true,
-            range_constraint: RangeConstraint::from_value(7.into()),
-        });
+        t.substitute_by_known(
+            &"B",
+            &SymbolicExpression::from_symbol("B", RangeConstraint::from_value(7.into())),
+        );
         assert_eq!(t.to_string(), "(A * 7) * X + 7 * Y");
-        t.apply_update(&VariableUpdate {
-            variable: "A",
-            known: true,
-            range_constraint: RangeConstraint::from_value(0.into()),
-        });
+        t.substitute_by_known(
+            &"A",
+            &SymbolicExpression::from_symbol("A", RangeConstraint::from_value(0.into())),
+        );
         assert_eq!(t.to_string(), "7 * Y");
     }
 
+    #[test]
+    fn substitute_known() {
+        let x = Qse::from_unknown_variable("X");
+        let y = Qse::from_unknown_variable("Y");
+        let a = Qse::from_known_symbol("A", RangeConstraint::default());
+        let b = Qse::from_known_symbol("B", RangeConstraint::default());
+        let mut t: Qse = (x * a + y) * b.clone() + b;
+        assert_eq!(t.to_string(), "(A * B) * X + B * Y + B");
+        t.substitute_by_known(
+            &"B",
+            &(SymbolicExpression::from_symbol("B", Default::default())
+                + SymbolicExpression::from(GoldilocksField::from(1))),
```

Comment:
This is not a realistic update, is it? Maybe replace `B` by `C + 1`, to avoid confusion?